src/GraphQL.Client/GraphQLHttpClient.cs

```diff
@@ -111,6 +116,62 @@ public IObservable<GraphQLResponse<TResponse>> CreateSubscriptionStream<TRespons
 
     #region Private Methods
 
+    private async Task<GraphQLHttpResponse<TResponse>> SendAPQHttpRequestAsync<TResponse>(GraphQLRequest request, CancellationToken cancellationToken = default)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        var savedQuery = request.Query;
+        bool useAPQ = false;
+
+        if (request.Query != null && _useAPQ && Options.EnableAutomaticPersistedQueries(request))
+        {
+            // https://www.apollographql.com/docs/react/api/link/persisted-queries/
+            const int APQ_SUPPORTED_VERSION = 1;
+            useAPQ = true;
+            // TODO: I suggest to change request.Extension type to public Dictionary<string, object?>?
+            //request.Extensions ??= new();
+            //request.Extensions["persistedQuery"] = new Dictionary<string, object>
+            //{
+            //    ["version"] = APQ_SUPPORTED_VERSION,
+            //    ["sha256Hash"] = Hash.Compute(request.Query),
+            //};
+            request.Extensions ??= new
+            {
+                version = APQ_SUPPORTED_VERSION,
+                sha256Hash = Hash.Compute(request.Query),
+            };
+            request.Query = null;
+        }
+
+        var response = await SendHttpRequestAsync<TResponse>(request, cancellationToken);
+
+        if (useAPQ && response.Errors?.Length > 0)
+        {
+            if (response.Errors.Any(error => string.Equals(error.Message, "PersistedQueryNotFound", StringComparison.CurrentCultureIgnoreCase)))
+            {
+                // GraphQL server supports APQ!
+
+                // Alas, for the first time we did not guess and in vain removed Query, so we return Query and
+                // send request again. This is one-time "cache miss", not so scary. By the way, after a second call,
+                // there is a probability of receiving PERSISTED_QUERY_NOT_FOUND error again, but it is so ghostly
+                // that this can and should be neglect so not to complicate the code without the need.
+                request.Query = savedQuery;
+                return await SendHttpRequestAsync<TResponse>(request, cancellationToken);
+            }
+            else if (response.Errors.Any(error => string.Equals(error.Message, "PersistedQueryNotSupported") ||
+                     response.Errors.Any(error => string.Equals(error.Message, "GraphQL query is missing.")))) // GraphQL.NET specific error message
+            {
+                // GraphQL server either supports APQ of some other version, or does not support it at all.
+                // Send a request for the second time. This is better than returning an error. Let the client work with APQ disabled.
+                _useAPQ = false;
```

Comment:
it might be important to specify in the docs that graphql should be provisioned as singleton in such case, or this process would repeat for every request e.g. in asp case

Reply:
It's a general design question not coupled with APQ.
https://github.com/graphql-dotnet/graphql-client/issues/516
`GraphQLHttpClient` is designed to be singleton for endpoint. @rose-a would you like me to add such a note into README?

Reply:
Yep that wouldn't hurt